keps/sig-auth/0034-20190123-harden-default-discovery-bindings.md

```diff
@@ -0,0 +1,71 @@
+  - sig-api-machinery
+reviewers:
+  - "@liggitt"
+  - "@tallclair"
+approvers:
+  - "@liggitt"
+  - "@tallclair"
+editor: TBD
+creation-date: 2019-01-23
+last-updated: 2019-01-23
+status: provisional
+---
+
+# Harden Default RBAC Discovery ClusterRole(Binding)s
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a KEP and for highlighting any additional information provided beyond the standard KEP template.
+[Tools for generating][] a table of contents from markdown are available.
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+The aim of this change is to remove the `system:unauthenticated` subjects from the `system:discovery` and `system:basic-user` ClusterRoleBindings, while preserving unauthenticated access to a genuinely non-sensitive subset of the current APIs (e.g. `GET /healthz`, `GET /version`). This will improve the default privacy and security posture for new clusters.
+
+## Motivation
+
+One of the work items resulting from the [CVE-2018-1002105](https://github.com/kubernetes/kubernetes/issues/71411) post-mortem was to [investigate hardening the default RBAC discovery ClusterRoleBindings](https://github.com/kubernetes/kubernetes/issues/72115) (i.e. `system:discovery` and `system:basic-user`) to limit potential avenues for similar attack. Additionally, the fact that API extensions are exposed by the default discovery bindings is surprising to some and represents a potential privacy concern (e.g. `/apis/self-flying-cars.unicorn.vc/v1/`).
+
+### Goals
+
+Limit the set of APIs which allow for unauthenticated access by default.
+
+### Non-Goals
+
+Disabling unauthenticated and/or anonymous access to all APIs by default.
+
+## Proposal
+
+* Remove the `GET /healthz`, `/version` and `/version/` PolicyRules from the default [`system:discovery` ClusterRole](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L198-L208)
+* Remove the `system:unauthenticated` subject group from the default [`system:discovery` and `system:basic-user` ClusterRoleBindings](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L526-L527)
+* Create a new default ClusterRole which prominently, explicitly documents its intent to freely disclose information (e.g. `system:public-info-viewer`) and contains the `GET /healthz`, `/version` and `/version/` PolicyRules removed from `system:discovery`
+* Create a new default ClusterRoleBinding for the new ClusterRole, which grants the `system:(un)authenticated` subject groups access.
+
+### Risks and Mitigations
```

Comment:
enumerating the endpoints that would now be forbidden to an anonymous user by default, and the pros and cons, would be useful. 

Example:

`/openapi/v2` would not longer be available to anonymous users by default.
* pro: can't see all defined custom resource names/schemas by default
* con: consumers that need schema info (like editors) must allow authenticating to the cluster to fetch it (though to be compatible with all potential clusters, even ones that disable anonymous auth, such consumers already needed to allow authenticating)

Reply:
how's this?

Reply:
how's this?